.github/api_guidelines.md

```diff
@@ -8,10 +8,132 @@ in slack channel [#forum-acs-api-design](https://redhat-internal.slack.com/archi
+  **must not** import from the following packages:
+  - [/proto/storage](https://github.com/stackrox/stackrox/blob/master/proto/storage)
+  - [/proto/api/v1](https://github.com/stackrox/stackrox/blob/master/proto/api/v1)
+  - [/proto/api/v2](https://github.com/stackrox/stackrox/blob/master/proto/api/v2)
+- All services of the type `APIServiceWithCustomRoutes` **must not** import from the following packages:
+  - [/proto/storage](https://github.com/stackrox/stackrox/blob/master/proto/storage) 
+  - [/proto/internalapi](https://github.com/stackrox/stackrox/blob/master/proto/internalapi)
+- Structs used by APIs, such as those defined in [/proto/api/](https://github.com/stackrox/stackrox/blob/master/proto/api/), 
+[/proto/internalapi](https://github.com/stackrox/stackrox/blob/master/proto/internalapi), **must not** be 
+written to database.
+
+Exposing internal data structures, especially the ones representing and affecting database schema directly at 
+the API can pose security risks. By carefully selecting and limiting the data exposed via the API, we can prevent 
+unauthorized access and potential data breaches which can have compliance ramifications. 
+
+Furthermore, exposing unnecessary data in the API can lead to overfetching, where the client receives more data 
+than required. This inefficiency increases network traffic, consumes extra bandwidth, and impacts API performance. 
+By limiting API data to what is actually needed by clients, we optimize resource usage and improve response times. 
+
+Consider the following database/internal structures as an example data structure used in Central.
+
+```
+Deployment {
+  string id
+  string name
+  repeated KeyValue labels
+  repeated KeyValue annotations
+  repeated Container containers
+}
+  
+Container {
+  string name
+  string image_name
+  repeated Volume volumes
+}
+```
+
+Exposing the above data structure in the API will always lead to reading the all `Container` bytes from the database, 
+even if the user does not need them. Instead, design the API such that users can request the information 
+they need, as below:
+
+```
+// GetDeploymentRequest requests deployment information. Deployment metadata is requested by default.
+GetDeploymentRequest {
+  Options {
+    bool get_container_spec `json: "getContainerSpec"`
+  }
+}
+
+// Default response
+GetDeploymentResponse {
+  Metadata {
+    string name `json: "name"`
+    string id `json: "id"`
+  }
+}
+
+// If `"getContainerSpec": true`
+GetDeploymentResponse {
+  Metadata {
+    string name `json: "name"`
+    string id `json: "name"`
+    ...
+  }
+  ContainerSpec {
+    string name `json: "name"`
+    string image_name `json: "imageName"`
+    ...
+  }
+}
+```
+
+As APIs evolve over time, adding or removing data fields becomes inevitable. By limiting the exposure of internal 
+database structures, we decouple the API from the database schema, allowing us to modify the internal structures
+without breaking existing client applications.
```

Comment:
I feel like this is the statement I'd expect as the main reason for decoupling opposed to the ones mentioned in l.45 (while they do apply, IMO this is the main reason we should this).

Reply:
We cannot label one is more important than the other. The outcome of each of the three pillars may have a different priority for different users. For example, noisy API responses are a real issue for which we have evidence.

I encourage all developers to look at all aspects with sufficient importance when designing APIs. The three pillars mentioned in this section influence the development and maintenance of the APIs.

Reply:
Think my main point is to have them in one section and not spread across, but I suppose it's optional.

Reply:
The approach here is associate the rationale with an example. This is because not all engineers have experience in API design, data modeling, protobufs, etc.